src/service/builder.js

```diff
@@ -0,0 +1,236 @@
+import {getServiceForDoc, registerServiceBuilderForDoc} from '../service';
+import {hasNextNodeInDocumentOrder, isIframed} from '../dom';
+import {removeItem} from '../utils/array';
+
+const ID = 'builder';
+
+/** @implements {../service.Disposable} */
+export class Builder {
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc  */
+  constructor(ampdoc) {
+    /** @private @const */
+    this.ampdoc_ = ampdoc;
+
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.observer_ = new win.IntersectionObserver((e) => this.observed_(e), {
+      // Root bounds are not important, so we can use the `root:null` for a
+      // top-level window.
+      root: isIframed(win) ? win.document : null,
+      rootMargin: '250% 31.25%',
+      threshold: 0.001,
+    });
+
+    /** @private @const {!Map<!AmpElement, {asap: boolean, isIntersecting: boolean}>} */
+    this.targets_ = new Map();
+
+    /** @private {?Array<!AmpElement>} */
+    this.parsingTargets_ = [];
+
+    /** @private {boolean} */
+    ampdoc.whenReady().then(() => this.checkParsing_());
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() =>
+      this.docVisibilityChanged_()
+    );
+  }
+
+  /** @override */
+  dispose() {
+    this.observer_.disconnect();
+    this.targets_.clear();
+    if (this.visibilityUnlisten_) {
+      this.visibilityUnlisten_();
+      this.visibilityUnlisten_ = null;
+    }
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  scheduleAsap(target) {
+    this.targets_.set(target, {asap: true, isIntersecting: false});
+    this.waitParsing_(target);
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  schedule(target) {
+    if (this.targets_.has(target)) {
+      return;
+    }
+
+    if (target.deferredBuild()) {
+      this.targets_.set(target, {asap: false, isIntersecting: false});
+      this.observer_.observe(target);
+    } else {
+      this.targets_.set(target, {asap: false, isIntersecting: true});
+    }
+
+    this.waitParsing_(target);
+    this.signalScanReady_();
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  unschedule(target) {
+    if (!this.targets_.has(target)) {
+      return;
+    }
+
+    this.targets_.delete(target);
+
+    this.observer_.unobserve(target);
+
+    if (this.parsingTargets_) {
+      removeItem(this.parsingTargets_, target);
+      this.checkParsing_();
+    }
+  }
+
+  /** @private*/
+  signalScanReady_() {
+    if (!this.scheduledReady_) {
+      this.scheduledReady_ = true;
+      const {win} = this.ampdoc_;
+      win.setTimeout(() => {
+        // This signal mainly signifies that some of the elements have been
+        // discovered and scheduled.
+        this.ampdoc_.signals().signal(READY_SCAN_SIGNAL);
+      }, 50);
+    }
+  }
+
+  /** @private */
+  docVisibilityChanged_() {
+    const vs = this.ampdoc_.getVisibilityState();
+    if (vs != VisibilityState.PAUSED && vs != VisibilityState.INACTIVE) {
```

Comment:
Nit: to match the if statement in `maybeBuild_`
```suggestion
    if (vs = VisibilityState.VISIBLE || vs == VisibilityState.HIDDEN || vs == VisibilityState.PRERENDER) {
```